common/libs/framing/frame_detector.h

```diff
@@ -0,0 +1,206 @@
+ *
+ * RxBuffer is responsible for underlying reception and buffering of incoming
+ * data. Also RxBuffer provides callbacks described by RxListener interface. In
+ * particular:
+ * OnCharacterMatch is called when MARK character is received
+ * OnRxComplete is called when the requested number of bytes was received
+ * OnRxError is called when an error occurs in the underlying reception
+ * infrastructure
+ *
+ * FrameDetector is modeled as a Finite State Machine having states:
+ * LOST - FrameDetector is lost in the stream of incoming bytes and will
+ * interpret the next MARK byte as an end of the frame.
+ * WAIT_FOR_START_MARKER - FrameDetector is synced and is waiting for MARK byte
+ * that will denote the start of the frame RECEIVING_FRAME - FrameDetector is
+ * synced and will interpret the next MARK byte as an end of the frame and will
+ * interpret the bytes in RxBuffer as a frame.
+ *
+ * OnRxError will cause FrameDetector to transition to the LOST state, frame
+ * being received will be ignored.
+ * ...MbbbbbbEbbbbMMbbbbbbbM
+ *    ^-----------^ this frame will be ignored
+ *                 ^-------^ this frame will be received
+ *
+ * OnRxComplete should not happen in normal reception. When FrameEncoder orders
+ * a reception from RxBuffer, it specifies maximum length of the frame expected,
+ * also FrameEncoder restarts reception upon receiving OnCharMatch callback.
+ * Thus, receiving OnRxComplete callback means that we get no MARK characters in
+ * the stream because sender is sending garbage or a frame that is longer than
+ * the expected length.
+ *
+ * FrameDetector keeps a copy of the last successfully received frame and a flag
+ * denoting if the last received frame was read by the caller.
+ * */
+
+template <class RxBuffer, int FRAME_BUF_LEN>
+class FrameDetector : public RxListener {
+public:
+  enum class State {
+    LOST, // Frame detector does not know where it is in the stream of
+          // bytes coming into RxBuffer
+    WAIT_FOR_START_MARKER, // Frame detector is waiting for a frame start marker
+    RECEIVING_FRAME        // Frame detector is receiving a frame
+  };
+
+  FrameDetector(RxBuffer &t) : rx_buffer_(t){};
+
+  // Starts frame detector
+  [[nodiscard]] bool Begin() {
+    state_ = State::LOST;
+    frame_available_ = false;
+    return rx_buffer_.Begin(this);
+  }
+
+  // Callback method called when RxBuffer is full
+  void OnRxComplete() override {
+    // We should never reach the full read of rx buffer.
+    // If we get here, this means, there are no marker
+    // chars in the stream, so we are lost
+    state_ = State::LOST;
+    RestartRX();
+  }
+
+  // Callback method called when RxBuffer receives a marker character;
+  // Note, MARK character is already written to RxBuffer so all conditions
+  // evaluating rx_buffer_.ReceivedLength() take that into account
+  void OnCharacterMatch() override {
+    marker_count_++;
+
+    switch (state_) {
+    case State::LOST:
+      if (IsDataReceived()) {
```

Comment:
All the states dispatch between IsDataReceived/IsOnlyMarkers/else - so these are essentially a subclassification of OnCharacterMatch event. WDYT about restructuring the method a little like this:

```
void OnCharacterMatch() override {
  if (IsDataReceived()) OnCharacterMatchWithData();
  else if (IsOnlyMarkersReceived()) OnSingleMarkerReceived();
  else { state = LOST; RestartRX(); }
}
void OnCharacterMatchWithData() {
  switch(state)...
}
void OnSingleMarkerReceived() {
  switch(state) ...
}

Reply:
Yes, I have noticed that also during my last session. Did as you suggested, but not sure if this results in more readable code.